File: 2024/src/main/java/dpr/aoc2024/Day23.java
package dpr.aoc2024;

import dpr.commons.Day;
import dpr.commons.Pair;
import dpr.commons.Util;

import java.util.ArrayList;
import java.util.Collection;
import java.util.Collections;
import java.util.HashSet;
import java.util.List;
import java.util.Set;
import java.util.stream.Collectors;

public class Day23 implements Day {
    @Override
    public void execute() {
        var lines = Util.getNotEmptyLinesFromFile(dayNum(), "input.txt");
        Pair<Object, Object> solution1And2 = part1And2(lines);
        System.out.println(solution1And2.first());
        System.out.println(solution1And2.second());
    }

    @Override
    public int dayNum() {
        return 23;
    }

    Pair<Object, Object> part1And2(List<String> lines) {
        Set<String> edges = new HashSet<>();
        Set<Set<String>> connections = lines.stream().map(line -> {
            String[] parts = line.split("-");
            Set<String> con = Set.of(parts[0], parts[1]);
            edges.addAll(con);
            return con;
        }).collect(Collectors.toSet());
        Set<List<String>> lans = new HashSet<>();
        edges.forEach(edge -> {
//            System.out.println(edge);
            List<String> connected = connections.stream()
                    .filter(c -> c.contains(edge))
                    .flatMap(Collection::stream)
                    .filter(e -> !e.equals(edge))
                    .sorted()
                    .toList();
            for (int i = 0; i < connected.size() - 1; i++) {
                for (int j = i + 1; j < connected.size(); j++) {
                    String a = connected.get(i);
                    String b = connected.get(j);
//                    System.out.println("Checking " + a + " and " + b);
                    Set<String> pair = Set.of(a, b);
                    if (connections.contains(pair)) {
                        List<String> list = new ArrayList<>(pair);
                        list.add(edge);
                        Collections.sort(list);
                        lans.add(list);
//                        System.out.println("connected: " + list);
                    }
                }
            }
        });
        long part1 = lans.stream().filter(lan -> lan.stream().anyMatch(e -> e.startsWith("t"))).count();
        int best = Integer.MIN_VALUE;
        String bestName = "";
        while (!lans.isEmpty()) {
            Set<String> cur = lans.stream().limit(1).flatMap(Collection::stream).collect(Collectors.toSet());
//            lans.remove(cur);
            while (true) {
                int prevSize = cur.size();
                Set<List<String>> toRemove = new HashSet<>();
                lans.forEach(l -> {
                    Set<String> intersection = new HashSet<>(l);
                    Set<String> left = new HashSet<>(l);
                    intersection.retainAll(cur);
                    if (intersection.size() == 3) {
                        toRemove.add(l);
                    } else if (intersection.size() == 2) {
                        left.removeAll(intersection);
                        String toCheck = left.stream().findAny().get();
                        if (cur.stream().allMatch(c -> {
                            Set<String> con = Set.of(c, toCheck);
                            return connections.contains(con);
                        })) {
                            cur.add(toCheck);
                            toRemove.add(l);
                        }
                    }
                });
                lans.removeAll(toRemove);
                if (prevSize == cur.size()) {
                    break;
                }
            }
            int size = cur.size();
            if (size > best) {
                best = size;
                bestName = cur.stream().sorted().collect(Collectors.joining(","));
            }
            lans.removeIf(lan -> lan.stream().anyMatch(cur::contains));
            connections.removeIf(con -> con.stream().anyMatch(cur::contains));
        }
        return new Pair<>(part1, bestName);
    }
}

File: 2017/src/main/kotlin/dpr/aoc2017/Day23.kt
package dpr.aoc2017

import dpr.commons.Util

object Day23 {
    @JvmStatic
    fun main(args: Array<String>) = Util.measureTime {
        val input = Util.getNotEmptyLinesFromFile("/23/input.txt")
        println(part1(input))
        println(part2(input))
    }

    @JvmStatic
    fun part1(input: List<String>): Int {
        val actions = Day18.parseInput(input)
        val context = Day18.Context(0)
        var mul = 0
        while (context.cur < actions.size) {
            val action = actions[context.cur]
            if (action is Day18.Action.Mul) {
                ++mul
            }
            context.cur += action.apply(context).first
        }
        return mul
    }

    @JvmStatic
    fun part2(input: List<String>): Long {
        val actions = Day18.parseInput(input)
        val context = Day18.Context(0)
        context.reg.put("a", 1)
        while (context.cur < actions.size) {
            if (context.cur == 11) {
                val b = context.reg.get("b")
                val d = context.reg.get("d")
                val e = context.reg.get("e")
                val div = b / d
                if (b % d == 0L && div >= e && d * div == b) {
                    context.reg.put("f", 0)
                    context.cur = 24
                    continue
                }
                context.cur = 20
                continue
            }
            if (context.cur == 20) {
                var d = context.reg.get("d")
                val b = context.reg.get("b")
                while (d < b) {
                    ++d
                    if (b % d == 0L) {
                        context.reg.put("d", d)
                        break
                    }
                }
                ++context.cur
                continue
            }
            context.cur += actions[context.cur].apply(context).first
        }

        return context.reg.get("h")
    }
}
